8304074: [JMX] Add an approximation of total bytes allocated on the Java heap by the JVM

phohensee opened the pull request:
Please review this addition to com.sun.management.ThreadMXBean that returns the total number of bytes allocated on the Java heap since JVM launch by both terminated and live threads.

Because this PR adds a new interface method, I've updated the JMM_VERSION to 4, but would be happy to update it to 3_1 instead.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8307396](https://bugs.openjdk.org/browse/JDK-8307396) to be approved

### Issues
 * [JDK-8304074](https://bugs.openjdk.org/browse/JDK-8304074): [JMX] Add an approximation of total bytes allocated on the Java heap by the JVM
 * [JDK-8307396](https://bugs.openjdk.org/browse/JDK-8307396): [JMX] Add an approximation of total bytes allocated on the Java heap by the JVM (**CSR**)


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to [5f2f86bb](https://git.openjdk.org/jdk/pull/13814/files/5f2f86bb50b644fb589dd3da548b0ba8f14a2006)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13814/head:pull/13814` \
`$ git checkout pull/13814`

Update a local copy of the PR: \
`$ git checkout pull/13814` \
`$ git pull https://git.openjdk.org/jdk.git pull/13814/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13814`

View PR using the GUI difftool: \
`$ git pr show -t 13814`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13814.diff">https://git.openjdk.org/jdk/pull/13814.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13814#issuecomment-1535343762)

dholmes-ora commented on src/hotspot/share/services/management.cpp:
```diff
@@ -2095,7 +2095,25 @@ jlong Management::ticks_to_ms(jlong ticks) {
+    // the final result may be "too small" because a thread might be retired before
+    // the loop gets to it and thus not be counted. If, on the other hand and done
+    // here, exited_allocated_bytes is added after the loop, the final result might be
+    // "too large" because a thread might be counted twice, once in the loop and agsin
```
typo agsin

phohensee replied:
Fixed.

dholmes-ora commented on src/hotspot/share/services/management.cpp:
```diff
@@ -2095,7 +2095,25 @@ jlong Management::ticks_to_ms(jlong ticks) {
+
+// Gets the amount of memory allocated on the Java heap since JVM launch.
+JVM_ENTRY(jlong, jmm_GetAllThreadAllocatedMemory(JNIEnv *env))
+    // There is a race between threads that exit during the loop and calling
+    // exited_allocated_bytes. If the  result is initialized with exited_allocated_bytes,
```
If you want a stable and accurate value did you try holding the Threads_lock while you iterate the threads? Or do it as a safepoint VMop?

shipilev replied:
I agree we should strive to get the value as accurate as possible. I think for operational use at scale, we need to avoid doing safepoints. Holding a `ThreadLock` might also penalize other code that (ab)uses threading (we frequently see thousands of threads coming and going, don't ask).

But I have a fundamental question here: since SMR/TLH gives us a snapshot of currently live threads, and it also protects us from seeing an exiting thread in bad state (ultimately, a `delete`-d one), why can't we just trust its `cooked_allocated_bytes`, and avoid adding allocated bytes on exit path? If we cannot trust that, can we make it trustable while thread is protected by SMR/TLH?

phohensee replied:
I thought about doing those, but both would slow down the app/JVM and very likely introduce p99.9/p100 latency outliers that we'd rather not see just because we're sampling. Also,

1. The existing thread allocated bytes implementation isn't particularly accurate either, in the sense that reality quickly gets away from the method return values. That was a conscious decision to go with speed and efficiency over stability/accuracy, so I went the same way for this implementation.
2. Cloud services sample at fairly long intervals in order to avoid overhead: 1 minute is common and 5 minutes not unheard of. Stability/accuracy within short time frames is unneeded with such long sampling intervals.

phohensee replied:
Afaiu, SMR/TLH keeps a terminated thread's TLS accessible, but doesn't stop the termination process.

If we initialize result with exited_allocated_bytes, the "too small" possibility is still there. We still have a window between that initialization and the creation of the threads iterator where threads may terminate and thus not be included in the iteration. And new threads may become active during the iteration and not be included. As Aleksey observes, an advantage is that threads that terminate during the iteration will be counted only once and we don't have to check for terminated threads.

If we stick with adding exited_allocated_bytes to result after the iteration, Volker is correct that we shouldn't include terminated threads in the sum because their allocated bytes values will have been added to exited_allocated_bytes. We have the same undercount possibility that initializing result with exited_allocated_bytes has, so this (fixed) approach can result in "too small" also, but "too large" goes away.

Looks like both approaches are equivalent, so let's go with initializing result with exited_allocated_bytes because it avoids a comparison in the loop.

dholmes-ora replied:
> Afaiu, SMR/TLH keeps a terminated thread's TLS accessible, but doesn't stop the termination process.

Incorrect. A thread cannot complete the termination process if it is contained by a TLH - see ` ThreadsSMRSupport::smr_delete` and the call to `wait_until_not_protected`.

But not sure that helps with the zero-else-double accounting problem. Any read of the "total accumulated bytes written to date" value is racing with terminating threads.

phohensee replied:
Resolved, see below.

shipilev commented on src/java.management/share/classes/sun/management/ThreadImpl.java:
```diff
@@ -525,6 +532,7 @@ private static native void getThreadInfo1(long[] ids,
     private static native void getThreadUserCpuTime1(long[] ids, long[] result);
     private static native long getThreadAllocatedMemory0(long id);
     private static native void getThreadAllocatedMemory1(long[] ids, long[] result);
+    private static native long getThreadAllocatedMemory2();
```
We can call this one `getAllThreadAllocatedMemory`, which obviates the need for `2` as the suffix.

phohensee replied:
Fixed.

shipilev commented on src/jdk.management/share/classes/com/sun/management/ThreadMXBean.java:
```diff
@@ -121,7 +156,7 @@ public interface ThreadMXBean extends java.lang.management.ThreadMXBean {
      *
      * @return an approximation of the total memory allocated, in bytes, in
-     * heap memory for the current thread
+     * heap memory for the current thread,
```
I am not sure if typos changes in the public API requires a CSR (albeit trivial one). Maybe skip these updates?

phohensee replied:
Fixed.

shipilev commented on test/jdk/com/sun/management/ThreadMXBean/ThreadAllocatedMemory.java:
```diff
@@ -201,38 +206,102 @@ private static void testGetThreadsAllocatedBytes()
+
+        // baseline should be positive
+        Thread curThread = Thread.currentThread();
+        long cumulative_size = mbean.getAllThreadAllocatedBytes();
```
Java style for variables is camel-case, `cumulativeSize`.

phohensee replied:
Fixed.

shipilev commented on test/jdk/com/sun/management/ThreadMXBean/ThreadAllocatedMemoryArray.java:
```diff
@@ -118,18 +116,8 @@ public static void main(String[] argv)
-            System.out.println(threads[i].getName() +
-                " Previous allocated bytes = " + sizes[i] +
-                " Current allocated bytes = " + newSize);
+            checkResult(threads[i], sizes[i], sizes1[i]);
```
Since we are cleaning up the test anyway, can we / should we rename `sizes` -> `before`, `size1` -> `after`?

shipilev commented on test/jdk/com/sun/management/ThreadMXBean/ThreadAllocatedMemoryArray.java:
```diff
@@ -173,11 +160,25 @@ public static void main(String[] argv)
 
 
+    private static void checkResult(Thread curThread,
+                                    long prev_size, long curr_size) {
```
camelCase arguments.

phohensee replied:
Fixed.

shipilev commented on test/jdk/com/sun/management/ThreadMXBean/ThreadAllocatedMemory.java:
```diff
@@ -298,22 +367,22 @@ public void run() {
-                " ThreadAllocatedBytes  = " + size2);
+            System.out.println(getName() +
+                " ThreadAllocatedBytes before = " + size1 +
+                " ThreadAllocatedBytes after = " + size2);
 
             if (size1 > size2) {
                 throw new RuntimeException(getName() +
-                    " ThreadAllocatedBytes = " + size1 +
-                    " > ThreadAllocatedBytes = " + size2);
+                    " ThreadAllocatedBytes before = " + size1 +
+                    " > ThreadAllocatedBytes after = " + size2);
```
Is this replaceable with `checkResult(...)`?

phohensee replied:
Yes. Done.

shipilev commented on test/jdk/com/sun/management/ThreadMXBean/ThreadAllocatedMemory.java:
```diff
@@ -201,38 +206,102 @@ private static void testGetThreadsAllocatedBytes()
     }
 
-    private static void checkResult(Thread curThread,
+    private static long checkResult(Thread curThread,
```
There is another `checkResult` below? Should they be replaced by a single method?

phohensee replied:
No, there's only one checkResult method. I changed the result type from void to long.

shipilev commented on src/hotspot/share/include/jmm.h:
```diff
@@ -52,7 +52,9 @@ enum {
   JMM_VERSION_2   = 0x20020000, // JDK 10
   JMM_VERSION_3   = 0x20030000, // JDK 14
-  JMM_VERSION     = JMM_VERSION_3
+  JMM_VERSION_3_0 = 0x20030000,
```
What's `JMM_VERSION_3_0`?

phohensee replied:
Removed.

shipilev commented on src/hotspot/share/services/management.cpp:
```diff
@@ -2095,7 +2095,25 @@ jlong Management::ticks_to_ms(jlong ticks) {
+      jlong size = thread->cooked_allocated_bytes();
+      result += size;
+    }
+    return result + ThreadService::exited_allocated_bytes();;
```
Double `;;`.

phohensee replied:
Fixed.

shipilev reviewed (commented):
Some comments follow.

simonis reviewed (commented):
Looks good in general. Please find my comments inline.

phohensee commented on src/hotspot/share/services/threadService.hpp:
```diff
@@ -102,6 +106,11 @@ class ThreadService : public AllStatic {
 
+  static jlong exited_allocated_bytes()       { return _exited_allocated_bytes; }
+  static void incr_exited_allocated_bytes(jlong size) {
+    Atomic::add(&_exited_allocated_bytes, size);
```
Fixed.

phohensee commented on src/hotspot/share/include/jmm.h:
```diff
@@ -52,7 +52,9 @@ enum {
   JMM_VERSION_2   = 0x20020000, // JDK 10
   JMM_VERSION_3   = 0x20030000, // JDK 14
-  JMM_VERSION     = JMM_VERSION_3
+  JMM_VERSION_3_0 = 0x20030000,
```
Removed.

phohensee commented on src/hotspot/share/include/jmm.h:
```diff
@@ -316,7 +318,8 @@ typedef struct jmmInterface_1_ {
                                                   jstring flag_name,
                                                   jvalue  new_value);
-  void*        reserved6;
+  jlong        (JNICALL *GetAllThreadAllocatedMemory)
```
Done.

phohensee commented on src/hotspot/share/services/management.cpp:
```diff
@@ -2264,7 +2279,7 @@ const struct jmmInterface_1_ jmm_interface = {
   jmm_FindDeadlockedThreads,
   jmm_SetVMGlobal,
-  nullptr,
+  jmm_GetAllThreadAllocatedMemory,
```
Done.

dholmes-ora commented on src/hotspot/share/services/management.cpp:
```diff
@@ -2095,7 +2095,21 @@ jlong Management::ticks_to_ms(jlong ticks) {
+    // There's no check for terminated threads in the loop because their
+    // allocated byte counts aren't included in exited_allocated_bytes
+    // when result is initialized.
+    jlong result = ThreadService::exited_allocated_bytes();
+    for (JavaThreadIteratorWithHandle jtiwh; JavaThread *thread = jtiwh.next();) {
```
If you call `exited_allocated_bytes` whilst you have an active `ThreadsListHandle` then you at least ensure you don't miss accounting for threads that are just about to terminate.

simonis replied:
Do you mean something like:
```cpp
    JavaThreadIteratorWithHandle jtiwh;
    jlong result = ThreadService::exited_allocated_bytes();
    while (JavaThread *thread = jtiwh.next()) {
      ...
```
That would be fine for me. Otherwise I agree with the current compromise between accuracy and speed.

dholmes-ora replied:
Yes. Though I realize my comment needed to make clear that you need a thread to update the count after it has removed itself from the thread-list, for this to be an improvement.

phohensee replied:
Done as Volker suggests.

dholmes-ora commented on src/hotspot/share/services/threadService.cpp:
```diff
@@ -167,6 +170,7 @@ void ThreadService::remove_thread(JavaThread* thread, bool daemon) {
     // We did not get here via JavaThread::exit() so current_thread_exiting()
     // was not called, e.g., JavaThread::cleanup_failed_attach_current_thread().
     decrement_thread_counts(thread, daemon);
+    ThreadService::incr_exited_allocated_bytes(thread->cooked_allocated_bytes());
```
By doing this here you increase the likelihood of double-accounting for this thread. If you do this after the thread is no longer on any threads-list you may miss its contribution entirely, but you won't double-count it.

dcubed-ojdk replied:
If the `thread` is no longer on any ThreadsList, then it is not safe to access `thread`.

dholmes-ora replied:
@dcubed-ojdk This is the current thread acting on itself

simonis replied:
This method (i.e. `ThreadService::remove_thread()`) is called from `Threads::remove()` *after* the thread was removed from the thread list:
```cpp
void Threads::remove(JavaThread* p, bool is_daemon) {
  ...
    // Maintain fast thread list
    ThreadsSMRSupport::remove_thread(p);
    ...
    ThreadService::remove_thread(p, is_daemon);
```

But if we reach here from `JavaThread::cleanup_failed_attach_current_thread()` as the comment implies (`JavaThread::cleanup_failed_attach_current_thread()` calls `Threads::remove()` in the case of an attach failure), calling `ThreadService::incr_exited_allocated_bytes()` is probably irrelevant, because a thread which failed to attach can't allocate anyway.

phohensee replied:
Done.

dholmes-ora commented on src/jdk.management/share/classes/com/sun/management/ThreadMXBean.java:
```diff
@@ -105,6 +105,41 @@ public interface ThreadMXBean extends java.lang.management.ThreadMXBean {
+    /**
+     * Returns an approximation of the total amount of memory, in bytes,
+     * allocated in heap memory since the Java virtual machine was launched,
+     * including the amount allocated by terminated threads.
```
This "including ..." part seems redundant - it is the value allocated since JVM launch.

phohensee replied:
Removed.

dholmes-ora commented on src/hotspot/share/runtime/atomic.hpp:
```diff
@@ -307,8 +307,8 @@ class Atomic : AllStatic {
   // Platform-specific implementation of add.  Support for sizes of 4
-  // bytes and (if different) pointer size bytes are required.  The
-  // class must be default constructable, with these requirements:
+  // and 8 bytes are required.  The class must be default constructable,
```
Comment change seems unnecessary.

phohensee replied:
I found that I had to add 64-bit atomic add support on 32-bit platforms (working on it), so I changed the comment to match the similar cmpxchg one.

phohensee replied:
Reverted, since 64-bit atomic add is no longer needed.

dholmes-ora commented on src/jdk.management/share/classes/com/sun/management/ThreadMXBean.java:
```diff
@@ -106,15 +106,19 @@ public interface ThreadMXBean extends java.lang.management.ThreadMXBean {
      *
+     * @implSpec The default implementation throws UnsupportedOperationException
+     * if the Java virtual machine implementation does not support thread
+     * memory allocation measuremust, and otherwise acts as though thread
```
Typo: measuremust

phohensee replied:
Fixed.

dholmes-ora commented on src/hotspot/share/services/management.cpp:
```diff
@@ -2097,14 +2097,18 @@ jlong Management::ticks_to_ms(jlong ticks) {
-    // when result is initialized.
-    jlong result = ThreadService::exited_allocated_bytes();
-    for (JavaThreadIteratorWithHandle jtiwh; JavaThread *thread = jtiwh.next();) {
+    // when result is initialized. The iterator is created and initialized
+    // before the call to exited_allocated_bytes to ensure we don't miss
+    // accounting for threads that are just about to terminate.
```
Sorry this is only accurate if a thread updates `exited_allocated_bytes` after it removes itself from the Threads-list.

phohensee replied:
Fixed, see below.

dholmes-ora commented on src/hotspot/share/services/threadService.cpp:
```diff
@@ -217,6 +221,7 @@ void ThreadService::current_thread_exiting(JavaThread* jt, bool daemon) {
   assert(!jt->is_terminated() && jt->is_exiting(), "must be exiting");
 
   decrement_thread_counts(jt, daemon);
+  ThreadService::incr_exited_allocated_bytes(jt->cooked_allocated_bytes());
```
Again this is too soon. This should be deferred until after the thread has removed itself from the threads-list.

phohensee replied:
Done.

dholmes-ora commented:
Also note that the current placement of the `incr_exited_allocated_bytes` skips accounting for hidden threads because those ThreadService methods skip hidden threads. You need to do this update as one of the last actions of a terminating/detaching thread.

dholmes-ora commented on src/hotspot/share/services/management.cpp:
```diff
@@ -2098,17 +2094,32 @@ jlong Management::ticks_to_ms(jlong ticks) {
-    // when result is initialized. The iterator is created and initialized
-    // before the call to exited_allocated_bytes to ensure we don't miss
-    // accounting for threads that are just about to terminate.
+    // A thread increments exited_allocated_bytes after being removed from
+    // the threads list (in ThreadService::remove_thread), which means there's
+    // a race between threads that exit during the loop and reading
+    // exited_allocated_bytes. If result is initialized with exited_allocated_bytes,
+    // the final result may be "too small" because a thread might be removed
+    // from the list before the loop gets to it and thus not be counted. If,
+    // on the other hand, exited_allocated_bytes is added after the loop,
+    // the final result might be "too large" because a thread might be counted
+    // twice, once in the loop and agsin in exited_allocated_bytes if it's
+    // removed from the list after it's encountered in the loop but before
+    // adding exited_allocated_bytes.
```
This is not an accurate description.

Once we have the TLH we have a fixed set of threads and none of them can update exited_allocated_bytes. If we read exited_allocated_bytes while the TLH is held then the count can be short for two reasons:
1. Newly started threads
2. Terminating threads that already escaped the TLH but which hadn't updated exited_allocated_bytes by the time we read it.
AFAICS with the current logic there is no possibility of double-accounting.

dholmes-ora replied:
Sorry I thought you had made the change to only update exited_allocated_bytes after the thread removed itself from the threads-list. It really makes things so much simpler.

phohensee replied:
I made that change. The comment now reads:
    // A thread increments exited_allocated_bytes in ThreadService::remove_thread                                                                                                                                                                                                                                                                                           
    // only after it removes itself from the threads list, and once a TLH is                                                                                                                                                                                                                                                                                                
    // created, no thread it references can remove itself from the threads                                                                                                                                                                                                                                                                                                  
    // list, so none can update exited_allocated_bytes. We therefore initialize                                                                                                                                                                                                                                                                                             
    // result with exited_allocated_bytes after after we create the TLH so that                                                                                                                                                                                                                                                                                             
    // the final result can only be short due to (1) threads that start after                                                                                                                                                                                                                                                                                               
    // the TLH is created, or (2) terminating threads that escape TLH creation                                                                                                                                                                                                                                                                                              
    // and don't update exited_allocated_bytes before we initialize result.

dholmes-ora reviewed (changes requested):
Sorry to be a pain but can you split the Atomic changes out of this PR into their own issue please.

Though having said that, now that you call incr_exited_allocayed_bytes() where you do, it is protected by the Threads_lock so no atomic add is needed now.

simonis commented:
Looks good now except for the points already mentioned by @dholmes-ora .

phohensee commented on src/hotspot/share/services/#allocationSite.hpp#:
Removed.

phohensee commented on src/hotspot/share/services/.#allocationSite.hpp:
Removed.

phohensee commented on src/hotspot/share/services/threadService.hpp:
```diff
@@ -61,6 +61,10 @@ class ThreadService : public AllStatic {
   static PerfVariable* _daemon_threads_count;
 
+  // As could this...
+  // Number of heap bytes allocated by termianted threads.
```
Fixed.

phohensee commented:
Makes things very much simpler without needing 64-bit atomic add. Hopefully all cleaned up.

dholmes-ora commented on src/hotspot/share/services/threadService.hpp:
```diff
@@ -102,6 +106,12 @@ class ThreadService : public AllStatic {
   static jlong get_live_thread_count()        { return _atomic_threads_count; }
   static jlong get_daemon_thread_count()      { return _atomic_daemon_threads_count; }
 
+  static jlong exited_allocated_bytes()       { return _exited_allocated_bytes; }
+  static void incr_exited_allocated_bytes(jlong size) {
+    // No need for atomicity, method is called under the Threads_lock
+    _exited_allocated_bytes += size;
+  }
```
As there is a lock-free read however you need to use Atomic::load and Atomic::store - epsecially for 32-bit systems.

phohensee replied:
Changed. Not sure atomic load is needed.

phohensee replied:
Even though this version passed the pre-submit tests, which include the new versions of ThreadAllocatedBytes.java, running it more often on a host with many hardware threads resulted in failures where getTotalAllocatedBytes did not monotonically increase. I believe what happened was that without an acquire/release, the hardware can float earlier stores to _exited_allocated_bytes above later ones. Adding acquire/release fixed the issue.

dholmes-ora replied:
That should not be possible if the update is always under the lock. acquire/release control what other memory updates you see when you see "this" update - they have no affect on the update itself.

dholmes-ora commented on test/jdk/com/sun/management/ThreadMXBean/ThreadAllocatedMemory.java:
```diff
@@ -201,38 +206,102 @@ private static void testGetThreadsAllocatedBytes()
+        }
+
+        // start threads
+        done = false; done1 = false;
```
Nit: separate lines please

phohensee replied:
Fixed.

dholmes-ora commented on src/hotspot/share/services/threadService.hpp:
```diff
@@ -108,8 +108,13 @@ class ThreadService : public AllStatic {
 
   static jlong exited_allocated_bytes()       { return _exited_allocated_bytes; }
```
This needs to be an atomic::load too .

phohensee replied:
Done.

dholmes-ora commented on src/hotspot/share/services/threadService.hpp:
```diff
@@ -108,8 +108,13 @@ class ThreadService : public AllStatic {
+    // No need for an atomic add because called under the Threads_lock,
+    // but because _exited_allocated_bytes is read concurrently, need
+    // atomic store to avoid readers seeing a partial update. Also need
+    // acquire/release because the hardware might move a subsequent
+    // store to _exited_allocated_bytes above this one.
+    jlong old = Atomic::load_acquire(&_exited_allocated_bytes);
+    Atomic::release_store(&_exited_allocated_bytes, old + size);
```
First yes Atomic::load is not needed here due to the lock, it is needed in the accessor above - sorry that wasn't clear. The Atomic store is needed of course.

But the load_acquire is definitely not needed as there is an implicit acquire when the lock is acquired, so any load of the field here must see the last value written under the lock. The release_store is also not needed because a release is for ensuring visibility of previous stores of which there are none, and has no affect on the visibility of the store done by the `release_store`.

So lets imagine a really weakly-ordered piece of hardware where all of the writing threads are completely properly synchronized by the use of the lock (else the hw is broken), but a lock-free reader can potentially see any of those writes out-of-order.  I think the only way to guarantee seeing the most recent write would be to issue a full fence() before the load. But I would be very surprised if any of our hardware actually operates in such a fashion.

phohensee replied:
Thanks for the explanation. I reverted the acquire/release and the atomic load in incr_exited_allocated_bytes and added an atomic load in exited_allocated_bytes.

dholmes-ora reviewed (commented):
Updates look good - thanks.

But presumably you still have those test failures now the acquire/release has been removed again?

phohensee commented:
Nope, test failures went away with the latest revision. I'm at a loss on that one. Perhaps it was the atomic load.

dholmes-ora reviewed (commented):
I'm happy with the basic implementation of this now - thanks. I'll leave any issues about spec and JMM version for serviceability folk to approve.

dholmes-ora reviewed (approved):
I will click Approve to get rid of the "requested changes" indicator but please wait for serviceability reviews.

dholmes-ora commented on src/hotspot/share/services/management.cpp:
```diff
@@ -2095,7 +2091,25 @@ jlong Management::ticks_to_ms(jlong ticks) {
+    // created, no thread it references can remove itself from the threads
+    // list, so none can update exited_allocated_bytes. We therefore initialize
+    // result with exited_allocated_bytes after after we create the TLH so that
+    // the final result can only be short due to (1) threads that start after
+    // the TLH is created, or (2) terminating threads that escape TLH creation
+    // and don't update exited_allocated_bytes before we initialize result.
```
Okay this is why you can't do monotonicity check in the tests. Imagine we have 10 threads all live and we call this function and so tally up all 10 threads. The next time we call this 5 of the threads are in the process of terminating - they have escaped the TLH but not yet updated the exited_allocated_bytes - so now we only tally up 5 threads and so the total appears to be far less than last time.

kevinjwalls replied:
The tests and all apps might thank us for keeping last known value, and not returning anything lower, ensuring montonicity.

phohensee replied:
Fixed using a high water mark.

simonis replied:
A simple fix for guaranteeing monotonicity, would be to add another field (e.g. `max_allocated_bytes`) which keeps the last result returned by `getTotalThreadAllocatedBytes()` and the latter would then return the maximum of the current computation and `max_allocated_bytes`.

phohensee replied:
That's what I did. :) Thanks for your review!

mlchung commented on src/jdk.management/share/classes/com/sun/management/ThreadMXBean.java:
```diff
@@ -105,6 +105,42 @@ public interface ThreadMXBean extends java.lang.management.ThreadMXBean {
+     * delay between the time an object is allocated and the time its size is
+     * recorded.
+     *
+     * @implSpec The default implementation throws UnsupportedOperationException
```
```suggestion
     * @implSpec The default implementation throws {@code UnsupportedOperationException}
```

mlchung commented on test/jdk/com/sun/management/ThreadMXBean/ThreadAllocatedMemory.java:
```diff
@@ -152,7 +158,7 @@ private static void testGetThreadAllocatedBytes()
         try {
             curThread.join();
         } catch (InterruptedException e) {
```
should it just let `InterruptedException` be thrown rather than swallowing it?  This pattern appears multiple places in this test and ThreadAllocatedMemoryArray.java test.

phohensee replied:
The original implementation grabbed this code from, e.g., test/jdk/java/lang/management/ThreadMXBean/ThreadUserTime.java. In the latter, InterruptedException causes the test to fail, but for some reason now lost, ThreadAllocatedMemory.java doesn't do that, but ThreadAllocatedMemoryArray.java does.

mlchung replied:
Since you are on this file, it may be good to clean them up.

phohensee replied:
Done.

mlchung reviewed (commented):
The specification looks good.   A minor comment in the test.

phohensee commented:
Yes, it seems to have happened when I did a merge of this branch with jdk master, which latter must have been in a strange state. Unfortunately, git merge didn't complain at all, so I didn't check before pushing.

dholmes-ora commented:
@phohensee not clear if you saw my comment a couple of days ago about why the monotonic increasing check in the test can't work.

phohensee commented:
Yes, I saw your comment and have been thinking about how to fix it, but haven't come up with anything. Ramki (ysr) pointed out that walking the thread list using SMR is very expensive when there are 1000s of threads, and that there isn't much overhead to collecting per-thread statistics: a thread local counter is incremented when TLABs are retired or there are direct allocations of large objects in eden or the old gen. So, he suggested adding an atomic increment of a global counter whenever a thread local counter is incremented. Atomic operations have been completing in the LLC (i.e., on-chip) for some time now, so the overhead is pretty small. What do you think, shall I reimplement that way?

dholmes-ora commented:
A global allocation counter sounds like a potential bottleneck to me and everyone has to pay for it. I'd want to see broad performance measurements for that and hear what the GC folk have to say.

dholmes-ora commented:
A global counter would also need to resurrect Atomic::add(jlong) for all platforms.

dholmes-ora commented:
The monotonicity issue can be addressed by tracking the last total and never reporting a smaller value. (Much as we used to cache the current time to avoid it going backwards.)

phohensee commented:
I added a high water mark to getTotalThreadAllocatedBytes (under a lock, lest we have a monotonicity issue on the high water mark), and updated the tests per Mandy's request.

mlchung commented:
Thanks for updating the test.   The spec looks good.   I'll leave the hotspot implementation for David and others to approve.

phohensee commented:
Thanks for your review, Mandy.

dholmes-ora commented on test/jdk/com/sun/management/ThreadMXBean/ThreadAllocatedMemory.java:
```diff
@@ -300,11 +302,16 @@ private static long checkResult(Thread curThread,
         return currSize;
     }
 
+    private static void reportUnexpected(Exception e, String reason) {
```
nit: 'when' or 'where' seems a more appropriate name than 'reason'

phohensee replied:
Changed to "when".

dholmes-ora reviewed (commented):
High watermark changes seem fine in principle but not sure about borrowing the `MonitoringSupport_lock ` to do the locking. At a minimum this comment should be expanded:
```
./share/runtime/mutexLocker.hpp:extern Mutex*   MonitoringSupport_lock;          // Protects updates to the serviceability memory pools.
```

phohensee commented:
Changed the MonitoringSupport_lock comment to "Protects updates to the serviceability memory pools and allocated memory high water mark". The lock was introduced to serialize updates to the G1 memory pool stats, but given that doesn't happen very often, and the name, imo it's reasonable to use it in a more general fashion.

dholmes-ora reviewed (approved):
Nothing further from me. Thanks.

phohensee commented:
Thanks for your in-depth review!

phohensee commented:
/integrate
